### DISCUSSION TITLE
How to find a "FieldRead" expression without "this." in class internal functions on Java.

### DISCUSSION THREAD

**Discussion by ZQ-Struggle** [github/codeql#12790] [Q&A] [answered]
How to identify a 'FieldRead' expression without the 'this' keyword in a Java class's internal functions? I am attempting to locate a 'FieldRead' expression within an internal function of a class. However, during coding, the 'this' keyword was omitted, as  'fruitMap' shows in the following figure.
![image](https://user-images.githubusercontent.com/34684672/230763789-76a3c0fd-73e2-4a7c-acdb-bde0d203852c.png)

If the source code uses the expression "this.fruitMap", it is easy to locate the corresponding field read. However, I am not sure how to identify the field read when "this" is omitted.
Please tell me the correct way to do so!!!

**Comment by Aurora-yzj** [comment]
I've come across a similar problem. Hope somebody could help with this issue, I'll be really appreciated!

**Comment by ZQ-Struggle** [comment]
I find a solution from this issue. It works, but I'm not sure it is the right way.
https://github.com/github/codeql/issues/10980
The key point is that the predicate 'getQualifier' does not consider the FieldRead expression without the 'this' keyword. However, 'getFieldQualifier' in 'semmle.code.java.dataflow.internal.DataFlowNodes::Public' provides such a feature. It utilizes the 'ImplicitInstanceAccess' class to locate the FieldRead that implicitly accesses an internal property.
It is a solution learned from another question. If anyone can provide a better solution, I would be grateful.

**Reply by atorralba** [reply]
If you want to access the implicit qualifier of the field access for dataflow purposes (e.g. find accesses to fields of a certain type), this is the correct way of doing it. If you just want to access the field read, see @jketema answer below.

**Reply by ZQ-Struggle** [reply]
Thank you! Actually, both of two expression is essential to me.

**Comment by jketema** [comment, marked as answer]
Hi @ZQ-Struggle,

Thanks for your question. To locate all field reads without a qualifier (i.e., with an implicit `this`). You can do the following:
```codeql
import java

from FieldRead v
where not exists(v.getQualifier())
select v
```

**Reply by ZQ-Struggle** [reply]
Thanks a lot for illustration.
